### DISCUSSION TITLE
How to model ManyToMany relationship between 2+ equal entities?

### DISCUSSION THREAD

**Discussion by hayksaryan** [mikro-orm/mikro-orm#4404] [Q&A] [answered]
Here are 2 quotes from the Doctrine documentation (which I guess also apply to MikroORM)
[Link to the quote](https://www.doctrine-project.org/projects/doctrine-orm/en/2.15/reference/unitofwork-associations.html#important-concepts)
> Changes made only to the inverse side of an association are ignored. Make sure to update both sides of a bidirectional association (or at least the owning side, from Doctrine's point of view)

> The owning side of a bidirectional association is the side Doctrine looks at when determining the state of the association, and consequently whether there is anything to do to update the association in the database.

Here's how I interpret it:

When the Author is the owner, this will work, all Book entities will be attached to the Author (id = 1) entity
```
const author = authorRepository.findOne({ where: { id: 1 } });
const books = bookRepository.findAll();

author.add(books);

await authorRepository.flush();
```

But this won't work

```
const book = bookRepository.findOne({ where: { id: 1 } });
const authors = authorRepository.findAll();

book.add(authors);

await bookRepository.flush();
```
So, specifying the `owner: true` means "this entity is responsible for adding/updating the relations".

Now back to my use case. Let's say I want to define biderectional ManyToMany relationships between Apple, Cherry, and Banana entities. These are all (interrelated) fruits, but different enough to be separate entities. I want the logic described above to work on both sides. In other words, not have owning/inverse sides, or rather have both sides be owners. 

How would I acheive that with MikroORM?

**Comment by B4nan** [comment]
It should work, MikroORM [propagates changes automatically](https://mikro-orm.io/docs/propagation) to the owning side, so as long as you add initialized entities, it is fine. This is different from doctrine. Adding not initialized references to M:N inverse side will silently fail, which is something I want to take a closer look in v6, as I think we could find a way to support that too.

> In other words, not have owning/inverse sides, or rather have both sides be owners.

You can't do that. And I didn't really understand why you would want that. Maybe provide a failing repro?

**Reply by hayksaryan** [reply]
Thanks for the quick response. 

Does this mean I should just set `owner: true` on an arbitrary entity?

**Reply by hayksaryan** [reply]
Apple, Cherry and Banana are an example (not really good), but they just reflect the equality of the relationship. All 3 fruits are interconnected. Between Apple and Cherry, none is the owner, same goes for all other combination of relations between 3 fruits. I will both do `apple.add(cherries)` and `cherry.add(apples)` in my codebase. If I set the Apple side as the owner, it feels natural to only do `apple.add(cherries)`. But setting an owner doesn't really make sense in the context of this relationship.

**Reply by hayksaryan** [reply]
It's also possible I might already need a graph database to support this kind of relationship at this point. But I'd rather not cross that bridge now.

**Reply by B4nan** [reply]
I think the real question is how your schema looks like, then we can talk how you can model the entities based on it.

**Reply by hayksaryan** [reply]
Will this suffice?
![image](https://github.com/mikro-orm/mikro-orm/assets/52959001/1d2661af-acad-44e4-8613-e0394f1c7424)

**Reply by B4nan** [reply, marked as answer]
I see, yeah that should be fine, from the ORM standpoint, the owner is the first FK in the pivot table, inverse is the second one, and as well as you have both sides initialized, you can work with the inverse just fine.

**Reply by hayksaryan** [reply]
So, to recap, in this specific use case, the owner for each relation can be either of the entities. Right?

________________
Also, the related part of the doc is a bit confusing. 

> Although this propagation works also for M:N inverse side, you should always use owning side to manipulate the collection.

Why so?


Really appreciate the help! :bow:

**Reply by B4nan** [reply]
Mainly because of the limitation I mentioned above with not initialized entities.
